ambry-store/src/main/java/com.github.ambry.store/DiskSpaceAllocator.java

```diff
@@ -0,0 +1,335 @@
+    File fileSizeDir = prepareDirectory(new File(reserveDir, Long.toString(sizeInBytes)));
+    File reserveFile;
+    do {
+      reserveFile = new File(fileSizeDir, generateFilename());
+    } while (!reserveFile.createNewFile());
+    Utils.preAllocateFileIfNeeded(reserveFile, sizeInBytes);
+    return reserveFile;
+  }
+
+  /**
+   * @param requirementsList the list of {@link DiskSpaceRequirements} objects to be accumulated
+   * @return a {@link Map} from segment sizes to the number of reserve segments needed at that size.
+   */
+  private Map<Long, Long> getOverallRequirements(Collection<DiskSpaceRequirements> requirementsList) {
+    Map<Long, Long> overallRequirements = new HashMap<>();
+    Map<Long, Long> swapSegmentsUsed = new HashMap<>();
+
+    for (DiskSpaceRequirements requirements : requirementsList) {
+      long segmentSizeInBytes = requirements.getSegmentSizeInBytes();
+      overallRequirements.put(segmentSizeInBytes,
+          overallRequirements.getOrDefault(segmentSizeInBytes, 0L) + requirements.getSegmentsNeeded());
+      swapSegmentsUsed.put(segmentSizeInBytes,
+          swapSegmentsUsed.getOrDefault(segmentSizeInBytes, 0L) + requirements.getSwapSegmentsInUse());
+    }
+    for (Map.Entry<Long, Long> sizeAndSegmentsNeeded : overallRequirements.entrySet()) {
+      long sizeInBytes = sizeAndSegmentsNeeded.getKey();
+      long origSegmentsNeeded = sizeAndSegmentsNeeded.getValue();
+      // Ensure that swap segments are only added to the requirements if the number of swap segments allocated to
+      // stores is lower than the minimum required swap segments.
+      long swapSegmentsNeeded = Math.max(requiredSwapSegmentsPerSize - swapSegmentsUsed.get(sizeInBytes), 0L);
+      overallRequirements.put(sizeInBytes, origSegmentsNeeded + swapSegmentsNeeded);
+    }
+    return Collections.unmodifiableMap(overallRequirements);
+  }
+
+  /**
+   * Create a directory if it does not yet exist and check that it is accessible.
+   * @param dir The directory to possibly create.
+   * @return the passed-in directory
+   * @throws IOException if the specified directory could not be created/read, or if a non-directory file with the same
+   *                     name already exists.
+   */
+  private static File prepareDirectory(File dir) throws IOException {
+    if (!dir.exists()) {
+      dir.mkdir();
+    }
+    if (!dir.isDirectory()) {
+      throw new IOException("Directory could not be created or is not a directory: " + dir.getAbsolutePath());
+    }
+    return dir;
+  }
+
+  /**
+   * @return a filename for a reserve file. This is reserve_ followed by a random UUID.
+   */
+  private static String generateFilename() {
+    return RESERVE_FILE_PREFIX + UUID.randomUUID();
+  }
+
+  private static class ReserveFileMap {
+    private final Map<Long, Queue<File>> internalMap = new HashMap<>();
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
```

Comment:
I've considered this. It was a bit worrisome given that there are nested data structure. 

I think the case that I was worried about is when one actor is adding a reserve file and another is removing a reserve file for a file size with one file currently in the queue. I thought that the step where we can remove an empty queue in `remove()` could remove a queue that was just appended to by `add()`. I suppose a solution might be to let empty queues stay in the data structure indefinitely. 

The other issue was the initialization lock, but for our use case we might be able to assume that no one will use the diskspaceallocator between BlobStore.start() and initializePool (given that compactionManager is started after DSA).

@vgkholla Do these compromises sound alright?